counter/src/main/java/org/infinispan/counter/CounterManager.java

```diff
@@ -0,0 +1,138 @@
+package org.infinispan.counter;
+
+import org.infinispan.Cache;
+import org.infinispan.counter.impl.atomic.AtomicCounter;
+import org.infinispan.counter.impl.atomic.ThresholdAtomicCounter;
+import org.infinispan.counter.impl.weak.WeakInfinispanCounter;
+import org.infinispan.factories.GlobalComponentRegistry;
+import org.infinispan.manager.EmbeddedCacheManager;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * The {@link CounterManager} creates and returns existing {@link Counter}s.
+ * <p>
+ * Each {@link CounterManager} is associated to a {@link EmbeddedCacheManager}. Internally, a {@link Cache} is used to
+ * distribute and provide concurrent control access to a {@link Counter} from multiples accesses from all the nodes.
+ * <p>
+ * It is thread-safe in the way that multiples threads can retrieve/create {@link Counter}s. If it is the first time a
+ * {@link Counter} is created, concurrent threads may block until it is properly initialized.
+ *
+ * @author Pedro Ruivo
+ * @since 9.0
+ */
+public class CounterManager {
+
+   public static final String COUNTER_CACHE_NAME = "__counters__";
```

Comment:
As we add more of these "system" caches, we should really put the names in a single place (and use the same naming convention)